File: solutions/csharp/complex-numbers/1/ComplexNumbers.cs
using System;

public readonly struct ComplexNumber
{
	private readonly double _real;
	private readonly double _imaginary;
	public ComplexNumber(double real, double imaginary)
	{
		_real = real;
		_imaginary = imaginary;
	}

	public double Real() =>
		_real;

	public double Imaginary() =>
		_imaginary;

	public ComplexNumber Mul(ComplexNumber other)
	{
		double realPart = (_real * other._real) - (_imaginary * other._imaginary);
		double imaginaryPart = (_real * other._imaginary) + (_imaginary * other._real);
		return new ComplexNumber(realPart, imaginaryPart);
	}

	public ComplexNumber Add(ComplexNumber other)
	{
		double realPart = _real + other._real;
		double imaginaryPart = _imaginary + other._imaginary;
		return new ComplexNumber(realPart, imaginaryPart);
	}

	public ComplexNumber Sub(ComplexNumber other)
	{
		double realPart = _real - other._real;
		double imaginaryPart = _imaginary - other._imaginary;
		return new ComplexNumber(realPart, imaginaryPart);
	}

	public ComplexNumber Div(ComplexNumber other)
	{
		double numerator1 = (_real * other._real) + (_imaginary * other._imaginary);
		double denominator = other._imaginary.Squared() + other._real.Squared();
		double realPart = numerator1 / denominator;

		double numerator2 = (_imaginary * other._real) - (_real * other._imaginary);
		double imaginaryPart = numerator2 / denominator;
		return new ComplexNumber(realPart, imaginaryPart);
	}

	public double Abs() =>
		Math.Sqrt(_real.Squared() + _imaginary.Squared());

	public ComplexNumber Conjugate() =>
		new(_real, -_imaginary);

	public ComplexNumber Exp()
	{
		ComplexNumber a = new(Math.Exp(_real), 0);
		ComplexNumber b = new(Math.Cos(_imaginary), 0);
		ComplexNumber c = new(0, Math.Sin(_imaginary));
		return a.Mul(b.Add(c));
	}

	public static implicit operator ComplexNumber(int integer) =>
		new(integer, 0);
}

public static class DoubleExtensions
{
	public static double Squared(this double value)
		=> Math.Pow(value, 2);
}

File: solutions/erlang/complex-numbers/1/src/complex_numbers.erl
-module(complex_numbers).

-export([abs/1, add/2, conjugate/1, divide/2, equal/2, exp/1, imaginary/1, mul/2, new/2,
  real/1, sub/2]).

-record(complex_number, {real_part, imaginary_part}).

-tolerance(0.005).

abs({complex_number, RealPart, ImaginaryPart1}) ->
  math:sqrt(math:pow(RealPart, 2) + math:pow(ImaginaryPart1, 2)).

add(
    {complex_number, RealPart1, ImaginaryPart1},
    {complex_number, RealPart2, ImaginaryPart2}
) ->
  new(RealPart1 + RealPart2, ImaginaryPart1 + ImaginaryPart2).

conjugate({complex_number, RealPart, ImaginaryPart}) ->
  new(RealPart, -ImaginaryPart).

divide(
    {complex_number, RealPart1, ImaginaryPart1},
    {complex_number, RealPart2, ImaginaryPart2}
) ->
  new(
    (RealPart1 * RealPart2 + ImaginaryPart1 * ImaginaryPart2) / (math:pow(RealPart2, 2) + math:pow(ImaginaryPart2, 2)),
    (RealPart2 * ImaginaryPart1 - RealPart1 * ImaginaryPart2) / (math:pow(RealPart2, 2) + math:pow(ImaginaryPart2, 2))
  ).

equal(
    {complex_number, RealPart1, ImaginaryPart1},
    {complex_number, RealPart2, ImaginaryPart2}
) ->
  {complex_number, RealPartDifference, ImaginaryDifference} =
    new(RealPart1 - RealPart2, ImaginaryPart1 - ImaginaryPart2),
  RealPartDifference < tolerance andalso ImaginaryDifference < tolerance.

exp({complex_number, _RealPart, ImaginaryPart}) ->
  new(math:cos(ImaginaryPart), math:sin(ImaginaryPart)).

imaginary({complex_number, _RealPart, ImaginaryPart}) ->
  ImaginaryPart.

mul(
    {complex_number, RealPart1, ImaginaryPart1},
    {complex_number, RealPart2, ImaginaryPart2}
) ->
  new(
    RealPart1 * RealPart2 - ImaginaryPart1 * ImaginaryPart2,
    RealPart2 * ImaginaryPart1 + RealPart1 * ImaginaryPart2
  ).

new(RealPart, ImaginaryPart) ->
  #complex_number{
    real_part = RealPart,
    imaginary_part = ImaginaryPart
  }.

real({complex_number, RealPart, _ImaginaryPart}) ->
  RealPart.

sub(
    {complex_number, RealPart1, ImaginaryPart1},
    {complex_number, RealPart2, ImaginaryPart2}
) ->
  new(RealPart1 - RealPart2, ImaginaryPart1 - ImaginaryPart2).
